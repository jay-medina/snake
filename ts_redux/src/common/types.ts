import { MapStateToProps as MSTP } from 'react-redux';
import { ThunkDispatch, ThunkAction } from 'redux-thunk';
import { AppAction } from '../store/actions';
import { Store } from 'redux';

export type GameThunkAction = ThunkAction<void, AppState, {}, AppAction>;

export type GameThunkDispatch = ThunkDispatch<AppState, {}, AppAction>;

export type DTP<DispatchProps, OwnProps = {}> = (
  dispatch: GameThunkDispatch,
  ownProps: OwnProps,
) => DispatchProps;

export type STP<StateProps, OwnProps = {}> = MSTP<StateProps, OwnProps, AppState>;

export type GameStore = Store<AppState, AppAction> & {
  dispatch: GameThunkDispatch;
};

export type GameState = 'Start' | 'Run' | 'GameOver';

export interface AppState {
  gameState: GameState;
  apple: Apple;
  snake: Snake;
  dimensions: {
    rows: number;
    cols: number;
  };
  score: Score;
}

export interface GridItem {
  row: number;
  col: number;
}

export type Apple = GridItem;

export type Direction = 'left' | 'right' | 'up' | 'down';

export interface Score {
  current: number;
  high: number;
}

export interface Snake {
  body: GridItem[];
  incrementTimer: number;
  lastTimestamp: number;
  direction: Direction;
  newDirection?: Direction;
}

export type TimeStamp = number;
