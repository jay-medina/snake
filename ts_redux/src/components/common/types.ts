export interface GridItem {
  row: number;
  col: number;
}

export type Apple = GridItem;

export type Snake = GridItem[];

export type GameState = 'start' | 'run' | 'gameover';
