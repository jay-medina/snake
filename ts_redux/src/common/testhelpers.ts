import { AppState, GameThunkDispatch, Snake } from './types'
import createMockStore, { MockStoreCreator, MockStoreEnhanced } from 'redux-mock-store'
import thunk from 'redux-thunk'

export type StoreCreator = MockStoreCreator<AppState, GameThunkDispatch>

export type MockStore = MockStoreEnhanced<AppState, GameThunkDispatch>

export const createTestStore = (): StoreCreator => createMockStore([thunk])

export const createMockSnake = (): Snake => ({
  body: [{ row: 5, col: 1 }, { row: 5, col: 2 }, { row: 5, col: 3 }],
  incrementTimer: 200,
  lastTimestamp: 100,
  direction: 'right',
})

export const createMockState = (): AppState => {
  return {
    gameState: 'Start',
    apple: { row: 1, col: 1 },
    snake: createMockSnake(),
    dimensions: {
      rows: 10,
      cols: 10,
    },
  }
}

export const mockRandom = (values: number[]) => {
  const mockMath = Object.create(global.Math)
  global.Math = mockMath

  const randomMock = jest.fn().mockReturnValue(0)

  values.forEach((value) => {
    randomMock.mockReturnValueOnce(value)
  })

  mockMath.random = randomMock
}
