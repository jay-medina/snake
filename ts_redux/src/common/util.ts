import { GridItem, Snake, AppState, Apple } from './types';

const isAtPosition = (gridItem1: GridItem) => (gridItem2: GridItem) => {
  return gridItem1.row === gridItem2.row && gridItem1.col === gridItem2.col;
};

const randomNumber = (num: number) => {
  return Math.floor(Math.random() * num);
};

const randomGridItem = (dimensions: AppState['dimensions']): GridItem => {
  return {
    row: randomNumber(dimensions.rows),
    col: randomNumber(dimensions.cols),
  };
};

export const isTheApple = isAtPosition;

export const isSnakeAtPosition = (snake: Snake['body']) => (gridItem: GridItem) =>
  !!snake.find(isAtPosition(gridItem));

export function findNewApplePosition(appState: AppState): Apple {
  const { snake, dimensions } = appState;

  let newApple = randomGridItem(dimensions);

  while (isSnakeAtPosition(snake.body)(newApple)) {
    newApple = randomGridItem(dimensions);
  }

  return newApple;
}

export function isSnakeDead(appState: AppState) {
  const { snake } = appState;
  return isSnakeInWall(appState) || isSnakeAtItself(snake.body);
}

function isSnakeInWall({ snake, dimensions }: AppState) {
  const [{ row, col }] = snake.body;
  const { rows, cols } = dimensions;

  return row < 0 || row >= rows || col < 0 || col >= cols;
}

function isSnakeAtItself(snake: Snake['body']) {
  const [head, ...rest] = snake;

  return isSnakeAtPosition(rest)(head);
}
