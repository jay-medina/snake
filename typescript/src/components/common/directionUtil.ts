import { Direction } from './types';

/**
 * Gets the next direction for the snake
 */
export function getNextDirection(currentDirection: Direction, newDirection: Direction) {
  if (isOppositeDirection(currentDirection, newDirection)) {
    return currentDirection;
  }

  return newDirection;
}

/**
 * Determines if the newDirection is opposite of the current direction
 *
 * @param {string} currentDirection
 * @param {string} newDirection
 */
export const isOppositeDirection = (currentDirection: Direction, newDirection: Direction) => {
  return (
    (currentDirection === 'left' && newDirection === 'right') ||
    (currentDirection === 'right' && newDirection === 'left') ||
    (currentDirection === 'up' && newDirection === 'down') ||
    (currentDirection === 'down' && newDirection === 'up')
  );
};
