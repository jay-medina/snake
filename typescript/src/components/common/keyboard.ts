import { Direction } from './types';

export const wireKeyboard = () => {
  let current: Direction = 'right';

  setUpKeyboardListeners((keyCode: number) => {
    const newDirection = mapArrowKeysToDirection(keyCode);

    if (newDirection) {
      current = newDirection;
    }
  });

  return {
    getDirection(): Direction {
      return current;
    },
    resetDirection() {
      current = 'right';
    },
  };
};

const LEFT = 37;
const UP = 38;
const RIGHT = 39;
const DOWN = 40;
const A = 65;
const W = 87;
const D = 68;
const S = 83;

/**
 * Maps the keycode to directions
 */
const mapArrowKeysToDirection = (keyCode: number): Direction | null => {
  switch (keyCode) {
    case LEFT:
    case A:
      return 'left';
    case UP:
    case W:
      return 'up';
    case RIGHT:
    case D:
      return 'right';
    case DOWN:
    case S:
      return 'down';
    default:
      return null;
  }
};

type NewDirectionCB = (keyCode: number) => void;

/**
 * Sets up the arrow key listener on the body element
 *
 */
const setUpKeyboardListeners = (newDirectionCb: NewDirectionCB) => {
  document.body.addEventListener('keydown', (e) => {
    newDirectionCb(e.keyCode);
  });
};
