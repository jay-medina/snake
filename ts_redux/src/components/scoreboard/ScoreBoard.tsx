import React from 'react'
import { connect } from 'react-redux'
import { STP } from '../../common/types'

export interface ScoreBoardProps {
  score: number
  highScore: number
}

interface ScoreProps {
  title: string
  num: number
}

const Score: React.StatelessComponent<ScoreProps> = ({ title, num }) => (
  <div className="snake__score">
    <div className="snake__score-title">{title}</div>
    <div className="snake__score-number">{num}</div>
  </div>
)

const ScoreBoard: React.SFC<ScoreBoardProps> = ({ score, highScore }) => (
  <div className="snake__scoreboard">
    <Score title="score" num={score} />
    <Score title="high" num={highScore} />
  </div>
)

ScoreBoard.displayName = 'Scoreboard'

const mSTP: STP<ScoreBoardProps> = ({ score }) => ({
  score: score.current,
  highScore: score.high,
})

export default connect(mSTP)(ScoreBoard)
