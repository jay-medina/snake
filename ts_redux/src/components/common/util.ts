import { GridItem, Snake } from 'store/util';

const isAtPosition = (gridItem1: GridItem) => (gridItem2: GridItem) => {
  return gridItem1.row === gridItem2.row && gridItem1.col === gridItem2.col;
};

export const isSnakeAtPosition = (snake: Snake) => (gridItem: GridItem) =>
  !!snake.find(isAtPosition(gridItem));

export const isTheApple = isAtPosition;
