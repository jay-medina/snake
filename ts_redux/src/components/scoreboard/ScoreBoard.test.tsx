import { mount } from 'enzyme';
import React from 'react';
import ScoreBoard from './ScoreBoard';
import { createTestStore } from '../../common/testhelpers';
import { Provider } from 'react-redux';

describe('<ScoreBoard /> ', () => {
  it('displays the scoreboard', () => {
    const store = createTestStore();

    const wrapper = mount(
      <Provider store={store}>
        <ScoreBoard />
      </Provider>,
    );
    expect(wrapper.find('Scoreboard')).toMatchSnapshot();
    wrapper.unmount();
  });
});
