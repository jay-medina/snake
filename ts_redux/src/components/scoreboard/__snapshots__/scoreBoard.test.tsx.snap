// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScoreBoard />  displays the scoreboard 1`] = `
<Scoreboard
  dispatch={[Function]}
  highScore={20}
  score={10}
>
  <div
    className="snake__scoreboard"
  >
    <Score
      num={10}
      title="score"
    >
      <div
        className="snake__score"
      >
        <div
          className="snake__score-title"
        >
          score
        </div>
        <div
          className="snake__score-number"
        >
          10
        </div>
      </div>
    </Score>
    <Score
      num={20}
      title="high"
    >
      <div
        className="snake__score"
      >
        <div
          className="snake__score-title"
        >
          high
        </div>
        <div
          className="snake__score-number"
        >
          20
        </div>
      </div>
    </Score>
  </div>
</Scoreboard>
`;
